File: service/ai_jobs/src/ai_jobs/tools/web_fetch.py
"""Versioned, model-facing webfetch contract and policy.

The worker logic is Rust; this is the thin definition the control plane needs to
admit a request, select the fixed profile, and validate what comes back. The
rules here are the same ones ``service/web_fetch/contracts`` enforces, and the
shared fixtures prove it.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from typing import Literal
from urllib.parse import urlsplit

from ai_jobs.contracts import (
    BrokerBudgets,
    ContractError,
    FailureCode,
    ToolFailure,
    require_exact_fields,
    require_object,
    require_string,
)
from ai_jobs.envelope import Envelope, ErrorCode, Inspection

TOOL_KEY = "web_fetch"
TOOL_VERSION = 1
WORKER_PROFILE = "web-fetch-v1"

MAX_URL_CHARS = 2_048
MAX_CONTENT_CHARS = 1_048_576
MAX_CONTENT_TYPE_CHARS = 200
#: Whole-run ceiling for both stages, inside the tool integration timeout.
RUN_DEADLINE_SECONDS = 25

Format = Literal["markdown", "text", "html"]
FORMATS: frozenset[str] = frozenset({"markdown", "text", "html"})


class ConversionMethod(StrEnum):
    NATIVE = "native"
    EXTRACTED = "extracted"
    BASIC = "basic"
    PASSTHROUGH = "passthrough"
    SOURCE = "source"


@dataclass(frozen=True, slots=True)
class WebFetchRequest:
    """A URL, and optionally an output format. Nothing else is accepted.

    There is deliberately no field for headers, cookies, credentials, rules, a
    timeout, a proxy, or a skip-scan flag; unknown fields are refused rather
    than ignored, so a caller cannot smuggle runtime settings past admission.
    """

    url: str
    format: Format = "markdown"

    @classmethod
    def from_dict(cls, value: object) -> "WebFetchRequest":
        data = require_object(value, "request")
        require_exact_fields(data, {"url"}, "request", {"format"})
        url = require_string(data["url"], "request.url", MAX_URL_CHARS)
        _check_fetchable(url)
        output = data.get("format", "markdown")
        if output not in FORMATS:
            raise ContractError("request.format must be markdown, text, or html")
        return cls(url=url, format=output)

    def to_dict(self) -> dict[str, object]:
        return {"url": self.url, "format": self.format}

    def storage_view(self) -> dict[str, object]:
        """What durable records may keep: a digest, never the URL itself."""
        return {"url_sha256": hashlib.sha256(self.url.encode()).hexdigest(), "format": self.format}


def _check_fetchable(url: str) -> None:
    parsed = urlsplit(url.strip())
    if parsed.scheme not in ("http", "https"):
        raise ContractError("request.url must be an HTTP(S) URL")
    if parsed.username or parsed.password:
        raise ContractError("request.url must not contain userinfo")
    if not parsed.hostname:
        raise ContractError("request.url must have a host")
    try:
        port = parsed.port
    except ValueError as exc:
        raise ContractError("request.url has an invalid port") from exc
    if port is not None and port not in (80, 443):
        raise ContractError("request.url must use the default web ports")


@dataclass(frozen=True, slots=True)
class WebFetchData:
    """The successful payload.

    Every field except ``source_id`` and the two enums came from the response,
    and the inspection stage scanned all of them - including the final URL and
    the media type - before this could be built.
    """

    source_id: str
    final_url: str
    http_status: int
    content_type: str
    retrieved_at: str
    format: Format
    conversion_method: ConversionMethod
    content: str

    @classmethod
    def from_dict(cls, value: object) -> "WebFetchData":
        data = require_object(value, "data")
        require_exact_fields(
            data,
            {
                "source_id",
                "final_url",
                "http_status",
                "content_type",
                "retrieved_at",
                "format",
                "conversion_method",
                "content",
            },
            "data",
        )
        source_id = require_string(data["source_id"], "data.source_id", 64)
        suffix = source_id.removeprefix("source-")
        if suffix == source_id or not suffix.isdigit():
            raise ContractError("data.source_id must be a run-scoped source identifier")

        final_url = require_string(data["final_url"], "data.final_url", MAX_URL_CHARS)
        parsed = urlsplit(final_url)
        if parsed.scheme not in ("http", "https") or not parsed.hostname or parsed.username or parsed.password:
            raise ContractError("data.final_url must be an HTTP(S) URL without userinfo")

        http_status = data["http_status"]
        if isinstance(http_status, bool) or not isinstance(http_status, int) or not 100 <= http_status <= 599:
            raise ContractError("data.http_status is not an HTTP status code")

        content_type = require_string(data["content_type"], "data.content_type", MAX_CONTENT_TYPE_CHARS)
        if not content_type.isascii() or any(ord(character) < 0x20 or ord(character) == 0x7F for character in content_type):
            raise ContractError("data.content_type must be printable ASCII")

        retrieved_at = require_string(data["retrieved_at"], "data.retrieved_at", 40)
        try:
            datetime.fromisoformat(retrieved_at.replace("Z", "+00:00"))
        except ValueError as exc:
            raise ContractError("data.retrieved_at must be an RFC 3339 timestamp") from exc

        output = data["format"]
        if output not in FORMATS:
            raise ContractError("data.format must be markdown, text, or html")
        try:
            method = ConversionMethod(require_string(data["conversion_method"], "data.conversion_method", 32))
        except ValueError as exc:
            raise ContractError("data.conversion_method is not a known conversion method") from exc

        content = require_string(data["content"], "data.content", MAX_CONTENT_CHARS, allow_empty=True)
        if output == "html" and method is not ConversionMethod.SOURCE:
            raise ContractError("html output is only ever returned as inert source")
        return cls(source_id, final_url, http_status, content_type, retrieved_at, output, method, content)

    def to_dict(self) -> dict[str, object]:
        return {
            "source_id": self.source_id,
            "final_url": self.final_url,
            "http_status": self.http_status,
            "content_type": self.content_type,
            "retrieved_at": self.retrieved_at,
            "format": self.format,
            "conversion_method": self.conversion_method.value,
            "content": self.content,
        }


def validate_envelope(value: object) -> Envelope:
    """Validate a `web_fetch:v1` envelope, payload included."""
    return Envelope.from_dict(
        value,
        tool=TOOL_KEY,
        tool_version=TOOL_VERSION,
        payload=WebFetchData.from_dict,
    )


def failure_envelope(code: ErrorCode, inspection: Inspection) -> Envelope:
    return Envelope.failure(TOOL_KEY, TOOL_VERSION, code, inspection)


class WebFetchDefinition:
    key = TOOL_KEY
    version = TOOL_VERSION
    worker_profile = WORKER_PROFILE

    def validate_request(self, value: object) -> WebFetchRequest:
        return WebFetchRequest.from_dict(value)

    def validate_result(self, value: object) -> Envelope:
        return validate_envelope(value)

    def grants_for(self, request: object) -> frozenset[str]:
        if not isinstance(request, WebFetchRequest):
            raise TypeError("expected WebFetchRequest")
        # One bounded GET. No search, no inference, no second fetch.
        return frozenset({"fetch"})

    def budgets_for(self, request: object) -> BrokerBudgets:
        if not isinstance(request, WebFetchRequest):
            raise TypeError("expected WebFetchRequest")
        return BrokerBudgets(
            search_calls=0,
            fetch_calls=1,
            fetched_bytes=1_048_576,
            input_tokens=0,
            output_tokens=0,
            elapsed_seconds=RUN_DEADLINE_SECONDS,
        )

    def storage_view(self, request: object) -> dict[str, object]:
        if not isinstance(request, WebFetchRequest):
            raise TypeError("expected WebFetchRequest")
        return request.storage_view()

    def public_failure(self, failure: ToolFailure) -> ToolFailure:
        """Clamp a control-plane failure to a fixed public one.

        A worker's own text never reaches a caller: anything that is not one of
        the codes this tool can produce becomes the generic internal failure.
        """
        allowed = {
            FailureCode.INVALID_REQUEST,
            FailureCode.TEMPORARILY_UNAVAILABLE,
            FailureCode.DEADLINE_EXCEEDED,
            FailureCode.UNSAFE_SOURCE,
        }
        if failure.code in allowed:
            return ToolFailure(code=failure.code, message=_PUBLIC_MESSAGES[failure.code])
        return ToolFailure(code=FailureCode.INTERNAL_ERROR, message=_PUBLIC_MESSAGES[FailureCode.INTERNAL_ERROR])


_PUBLIC_MESSAGES = {
    FailureCode.INVALID_REQUEST: "The request is not a valid web_fetch:v1 request.",
    FailureCode.TEMPORARILY_UNAVAILABLE: "The fetch capability is unavailable.",
    FailureCode.DEADLINE_EXCEEDED: "The request did not complete within its deadline.",
    FailureCode.UNSAFE_SOURCE: "The requested destination is not allowed.",
    FailureCode.INTERNAL_ERROR: "The tool could not complete this request.",
}

File: service/web_fetch/contracts/src/web_fetch.rs
//! The `web_fetch:v1` request and result payload.

use serde::{Deserialize, Serialize};
use time::OffsetDateTime;
use time::format_description::well_known::Rfc3339;
use url::Url;

use crate::envelope::{Envelope, Payload};
use crate::limits::{MAX_CONTENT_CHARS, MAX_CONTENT_TYPE_CHARS, MAX_URL_CHARS};
use crate::{ContractError, Result};

pub const TOOL: &str = "web_fetch";
pub const TOOL_VERSION: u32 = 1;

/// A validated webfetch envelope.
pub type WebFetchEnvelope = Envelope<WebFetchData>;

/// Output format. It selects deterministic conversion only; no value of it
/// changes what is inspected or how a match is handled.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Format {
    #[default]
    Markdown,
    Text,
    /// The decoded source, returned as inert text. It skips conversion, never
    /// scanning.
    Html,
}

/// Which converter produced the returned content. Fixed identifiers: a
/// converter reports what it is, and never returns a suggestion, a URL, or an
/// instruction to the caller.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConversionMethod {
    /// Native HTML cleanup and conversion.
    Native,
    /// Main-content extraction, then conversion.
    Extracted,
    /// Conversion without aggressive boilerplate removal.
    Basic,
    /// Non-HTML text returned as received after decoding.
    Passthrough,
    /// Decoded HTML source returned as inert text.
    Source,
}

/// What the model may supply: a URL, and optionally an output format.
///
/// There is deliberately nowhere to put headers, cookies, credentials, rules, a
/// timeout, a proxy, or a skip-scan flag — unknown fields are rejected outright.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WebFetchRequest {
    pub url: String,
    #[serde(default)]
    pub format: Format,
}

impl WebFetchRequest {
    pub fn new(url: impl Into<String>, format: Format) -> Self {
        Self {
            url: url.into(),
            format,
        }
    }

    pub fn from_json(value: &str) -> Result<Self> {
        let request: Self = serde_json::from_str(value)
            .map_err(|error| ContractError::new(format!("request is not valid: {error}")))?;
        request.validated_url()?;
        Ok(request)
    }

    /// Parse and check the URL, returning the exact form the fetcher will use.
    ///
    /// The fragment is dropped because it is never sent on the wire; everything
    /// else must already be acceptable rather than be repaired here.
    pub fn validated_url(&self) -> Result<Url> {
        let url = self.destination_url()?;
        check_fetchable(&url)?;
        Ok(url)
    }

    /// The same parse without the port rule, for the fetcher, which checks ports
    /// against its own destination policy on every hop including the first.
    pub fn destination_url(&self) -> Result<Url> {
        if self.url.len() > MAX_URL_CHARS {
            return Err(ContractError::new(format!(
                "request.url must be at most {MAX_URL_CHARS} characters"
            )));
        }
        let mut url = Url::parse(self.url.trim())
            .map_err(|error| ContractError::new(format!("request.url is not a URL: {error}")))?;
        check_destination(&url)?;
        url.set_fragment(None);
        Ok(url)
    }
}

/// Destination rules that hold for the requested URL and for every redirect
/// target, enforced again in the fetcher before each hop is attempted.
///
/// The port rule lives in [`check_fetchable`] rather than here because the
/// fetcher checks ports against its own destination policy, which is also what
/// decides which addresses are reachable at all.
pub fn check_destination(url: &Url) -> Result<()> {
    match url.scheme() {
        "http" | "https" => {}
        other => {
            return Err(ContractError::new(format!(
                "unsupported URL scheme: {other}"
            )));
        }
    }
    if !url.username().is_empty() || url.password().is_some() {
        return Err(ContractError::new("URL userinfo is not supported"));
    }
    if url.host_str().is_none_or(str::is_empty) {
        return Err(ContractError::new("URL must have a host"));
    }
    Ok(())
}

/// [`check_destination`], plus the v1 rule that only the default web ports are
/// fetchable.
pub fn check_fetchable(url: &Url) -> Result<()> {
    check_destination(url)?;
    match url.port() {
        None | Some(80 | 443) => Ok(()),
        Some(port) => Err(ContractError::new(format!("port {port} is not supported"))),
    }
}

/// The successful payload. Every string in it except the server-owned enums and
/// `source_id` came from the response, and is inspected before release.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WebFetchData {
    /// Run-scoped opaque identifier, not a handle to anything.
    pub source_id: String,
    pub final_url: String,
    pub http_status: u16,
    pub content_type: String,
    pub retrieved_at: String,
    pub format: Format,
    pub conversion_method: ConversionMethod,
    pub content: String,
}

impl Payload for WebFetchData {
    const TOOL: &'static str = TOOL;
    const TOOL_VERSION: u32 = TOOL_VERSION;

    fn validate(&self) -> Result<()> {
        if !is_source_id(&self.source_id) {
            return Err(ContractError::new(
                "data.source_id must be a run-scoped source identifier",
            ));
        }
        if self.final_url.len() > MAX_URL_CHARS {
            return Err(ContractError::new("data.final_url is too long"));
        }
        let final_url = Url::parse(&self.final_url)
            .map_err(|error| ContractError::new(format!("data.final_url is not a URL: {error}")))?;
        // Reporting, not authorizing: the port a response actually came from is
        // the fetcher's policy decision, already made. What matters here is that
        // the reported URL is a plain HTTP(S) URL with no userinfo in it.
        check_destination(&final_url)?;
        if !(100..=599).contains(&self.http_status) {
            return Err(ContractError::new(
                "data.http_status is not an HTTP status code",
            ));
        }
        if self.content_type.is_empty() || self.content_type.len() > MAX_CONTENT_TYPE_CHARS {
            return Err(ContractError::new(
                "data.content_type is not a usable media type",
            ));
        }
        if self
            .content_type
            .chars()
            .any(|character| character.is_control() || !character.is_ascii())
        {
            return Err(ContractError::new(
                "data.content_type must be printable ASCII",
            ));
        }
        OffsetDateTime::parse(&self.retrieved_at, &Rfc3339)
            .map_err(|_| ContractError::new("data.retrieved_at must be an RFC 3339 timestamp"))?;
        if self.content.chars().count() > MAX_CONTENT_CHARS {
            return Err(ContractError::new(
                "data.content exceeds the returned-content limit",
            ));
        }
        if self.format == Format::Html && self.conversion_method != ConversionMethod::Source {
            return Err(ContractError::new(
                "html output is only ever returned as inert source",
            ));
        }
        Ok(())
    }
}

fn is_source_id(value: &str) -> bool {
    value.strip_prefix("source-").is_some_and(|suffix| {
        !suffix.is_empty() && suffix.len() <= 32 && suffix.chars().all(|c| c.is_ascii_digit())
    })
}
